### PR TITLE
Update

### PR DESCRIPTION
**Pull request by AkbarSheikh-debug**
This PR introduces two major improvements to the data augmentation pipeline in augment.py:

Native Albumentations Keypoints & Segments Support Fixed the long-standing TODO where the Albumentations wrapper ignored segments (for instance segmentation) and keypoints (for pose estimation). Previously, using spatial transforms via Albumentations would result in misaligned masks or poses. The logic now actively routes them through an internally managed KeypointParams system, meaning spatial transforms will keep your masks and poses 100% aligned out of the box.

Native RandomErasing (Cutout) Augmentation Added a fully-native RandomErasing class directly into the YOLO preprocessing pipeline without having to rely on torchvision during detection/segmentation. This is hooked up cleanly to the erasing hyperparameter in v8_transforms. You can now easily enable Cutout-style box occlusions to improve model robustness simply by updating your training configs (e.g., setting erasing: 0.4).

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds a new `RandomErasing` augmentation and extends Albumentations support to preserve keypoints and segmentation points during spatial transforms. 🎨🧩

### 📊 Key Changes
- Introduces a new `RandomErasing` class in `ultralytics/data/augment.py` with configurable probability, erase area, aspect ratio, and fill value.
- Integrates `RandomErasing` into `v8_transforms()` using `hyp.erasing`, enabling optional patch-based image occlusion during training.
- Adds a dedicated `self.transform_kpt` Albumentations pipeline with bounding box and keypoint handling for spatial transforms.
- Updates Albumentations `__call__()` logic to transform and restore:
  - bounding boxes
  - pose keypoints
  - segmentation polygon points
- Preserves previous bbox-only behavior when no keypoints or segments are present.

### 🎯 Purpose & Impact
- Improves augmentation diversity and model robustness by simulating occlusion with random erased regions. 🛡️
- Expands Albumentations compatibility for pose and segmentation workflows so annotations stay aligned after transforms. 🎯
- Helps reduce annotation corruption risk for multi-task training pipelines that include boxes, keypoints, and segments.
- May benefit users training detection, pose, and segmentation models, though this change would especially benefit from validation tests around transformed keypoint and segment consistency.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @AkbarSheikh-debug, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you shortly. Please review the checklist below to help ensure a smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/data/augment.py [hunk 1]
```diff
@@ -1939,13 +1995,62 @@ def __call__(self, labels: dict[str, Any]) -> dict[str, Any]:
...
+                    kpts = []
+                    kpt_labels = []
+
+                    if keypoints is not None:
+                        kpts_flat = keypoints.reshape(-1, 3)[:, :2].tolist()
+                        kpts.extend(kpts_flat)
+                        kpt_labels.extend([0] * len(kpts_flat))
+
+                    if len(segments):
+                        seg_flat = np.concatenate(segments, axis=0) if isinstance(segments, (list, tuple)) else segments.reshape(-1, 2)
```
Comment (UltralyticsAssistant) [inline ultralytics/data/augment.py:2012]: ⚠️ **HIGH**: Flattening segmentation polygons and sending them through `keypoint_params` is not geometry-preserving for crop-like spatial transforms. Affine/flip is fine, but any crop/clip transform needs polygon clipping and may need new boundary vertices; transforming only the existing points can leave invalid or out-of-bounds polygons.

File: ultralytics/data/augment.py [hunk 2]
```diff
@@ -1939,13 +1995,62 @@ def __call__(self, labels: dict[str, Any]) -> dict[str, Any]:
...
+                        kpts.extend(seg_flat.tolist())
+                        kpt_labels.extend([1] * len(seg_flat))
+
+                    new = self.transform_kpt(image=im, bboxes=bboxes, class_labels=cls, keypoints=kpts, keypoint_labels=kpt_labels)
```
Comment (UltralyticsAssistant) [inline ultralytics/data/augment.py:2016]: ⚠️ **HIGH**: This loses the per-instance mapping between boxes and their keypoints/segments. Albumentations can drop a subset of `bboxes`, but the flattened `keypoints` list still contains points for every original instance, so after reconstruction `instances` can end up with fewer boxes than keypoint/segment entries and mismatched targets.

File: ultralytics/data/augment.py [hunk 3]
```diff
@@ -1939,13 +1995,62 @@ def __call__(self, labels: dict[str, Any]) -> dict[str, Any]:
...
+                            new_segments = []
+                            curr = 0
+                            for s in segments:
+                                new_segments.append(new_seg_flat[curr:curr+len(s)])
+                                curr += len(s)
+                            segments = new_segments
+                        else:
+                            segments = new_seg_flat.reshape(segments.shape)
+
+                    labels["instances"].update(bboxes=np.array(new["bboxes"], dtype=np.float32) if len(new["class_labels"]) else bboxes,
```
Comment (UltralyticsAssistant) [inline ultralytics/data/augment.py:2040]: ⚠️ **HIGH**: When `new["class_labels"]` is empty, this still writes transformed keypoints/segments back into `instances` while keeping the original image and original boxes/classes. That leaves annotations in the transformed coordinate frame attached to an untransformed image. The update of keypoints/segments needs the same success guard as the image/box update.

File: ultralytics/data/augment.py [hunk 4]
```diff
@@ -1887,6 +1934,15 @@ def __init__(self, p: float = 1.0, transforms: list | None = None) -> None:
                 if self.contains_spatial
                 else A.Compose(T)
             )
+            self.transform_kpt = (
```
Comment (UltralyticsAssistant) [inline ultralytics/data/augment.py:1937]: 💡 **MEDIUM**: `self.transform_kpt` never gets the deterministic seed that is applied to `self.transform`. With albumentations>=1.4.21 this means samples that go through the keypoint/segment path will ignore the caller's seed and produce non-reproducible augmentations.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The new `RandomErasing` addition looks straightforward, but the Albumentations keypoint/segment path has a few correctness issues. In particular, bbox filtering is not propagated back to the flattened keypoints/segments, zero-box results can desynchronize annotations from the image, and polygon handling via `keypoint_params` is not safe for crop-style transforms.

💬 Posted 4 inline comments

**Comment by glenn-jocher** [comment]
Thanks for tackling the long-standing Albumentations segments/keypoints TODO and proposing native random erasing, Akbar, we really appreciate the effort here. Unfortunately this PR bundles two independent features and, as written, wiring the `erasing` hyperparameter (documented in default.yaml as classification-only, default 0.4) into `v8_transforms` would silently change random-erasing behavior for every detection and segmentation training run, and the new class duplicates torchvision's existing RandomErasing already used in the classification pipeline. The branch has also gone stale with merge conflicts and no accompanying tests. We're going to close this for now, but we'd genuinely welcome a focused follow-up, ideally a single-concern PR for the Albumentations keypoints/segments support with a small regression test, and we'd be happy to review it. Thank you again for contributing to YOLO!
